### ISSUE TITLE
Call SetBasePath when IHostingEnvironment (or equivalent) is available

### ISSUE THREAD

**Issue by pintojoa** [Nexmo/nexmo-dotnet#17]
Following the instructions on the read me when sending one SMS I'm getting the following error:

An unhandled exception occurred while processing the request.
NullReferenceException: Object reference not set to an instance of an object.
Nexmo.Api.ConfigurationExtensions.KeyValueParser.AddToDictionary(XElement element, Stack<string> context, SortedDictionary<string, string> results)

TypeInitializationException: The type initializer for 'Nexmo.Api.Configuration' threw an exception.
Nexmo.Api.Request.ApiRequest.GetBaseUriFor(Type component, string url)

I'm using the VS Pro 2017RC with dotnetcore 1.1.0.

**Comment by pintojoa** [comment]
I've created a new settings file with the name "settings.json" on the root of the webapplication (at the same level of the appsettings.json file)

The content of the settings.json file is:
```
{
  "ConnectionStrings": {
    "DefaultConnection": "Server=(localdb)\\MSSQLLocalDB;Database=_CHANGE_ME;Trusted_Connection=True;MultipleActiveResultSets=true"
  },
  "appSettings": {
    "Nexmo.UserAgent": "myApp/1.0",
    "Nexmo.Url.Rest": "https://rest.nexmo.com",
    "Nexmo.Url.Api": "https://api.nexmo.com",
    "Nexmo.api_key": "api_key",
    "Nexmo.api_secret": `"api_secret",`
    "Nexmo.Application.Id": "ffffffff-ffff-ffff-ffff-ffffffffffff",
    "Nexmo.Application.Key": "c:\\path\\to\\your\\application\\private.key"
  }
}
```

**Comment by smithrobs** [comment]
Thank you. I pushed out v2.1.2 to address this issue.

The major issue is that the netcore webapp template includes a `web.config` file with contents similar to the following:

```
    <handlers>
      <add name="aspNetCore" path="*" verb="*" modules="AspNetCoreModule" resourceType="Unspecified"/>
    </handlers>
```

The code from MS that I used to support web.config files did not check for null when config elements contained the attribute `name` but not the attribute `connectionString`. I added this check as well as only pull settings from descendants of `<appSettings>` and `<connectionStrings>` elements.

I also added `appsettings.json` to the list of files that are searched for configuration. If you'd like, you could merge your `settings.json` into there.

Please let me know whether this release addresses this issue or not.

**Comment by pintojoa** [comment]
I've just update the nugget package. Now I'm only using the following appsettings.json.

```
{
  "Logging": {
    "IncludeScopes": false,
    "LogLevel": {
      "Default": "Debug",
      "System": "Information",
      "Microsoft": "Information"
    }
  },
  "appSettings": {
    "Nexmo.UserAgent": "myApp/1.0",
    "Nexmo.Url.Rest": "https://rest.nexmo.com",
    "Nexmo.Url.Api": "https://api.nexmo.com",
    "Nexmo.api_key": "key",
    "Nexmo.api_secret": "secret",
    "Nexmo.Application.Id": "ffffffff-ffff-ffff-ffff-ffffffffffff",
    "Nexmo.Application.Key": "c:\\path\\to\\your\\application\\private.key"
  }
}
```
But it seems that the configurations are being captured.

When I try to send one sms I get the error from the API "Missing username"

**Comment by smithrobs** [comment]
Thank you for your patience. Would you try this:
1. In Solution Explorer, right-click on `appsettings.json` and click on Properties.
2. Ensure "Copy to Output Directory" is set to "Copy always"

![i17](https://cloud.githubusercontent.com/assets/245836/20990359/4b856ac4-bca6-11e6-8da1-872fa1f8fcca.png)

**Comment by pintojoa** [comment]
It seems to be working now.

In our case would also be nice if we could pass the configuration throught the constructor to be able to change the api_key depending on the destination. Is this possible?

**Comment by smithrobs** [comment]
Thanks. I'm going to rename this and keep it open until the SetBasePath issue is resolved.

You want the ability to pass in additional api_key/api_secret pairs, correct? There is no clear way to do that at this time. You can however override it through the library configuration:

```Nexmo.Api.Configuration.Instance.Settings["appSettings:Nexmo.api_key"] = "override";```

I've opened #18 to track this.

**Comment by smithrobs** [comment]
@pintojoa I just pushed a pre-release version v2.2.0-rc1 to NuGet. This release includes the ability to specify API credentials on a per-request basis. The IHostingEnvironment issue is still outstanding, but it's one of my next priorities.

**Comment by sidsharma27** [comment]
@smithrobs having the configuration issue with 2.2.0-rc2 as I was trying to get around [this error](https://github.com/Nexmo/nexmo-dotnet/issues/22) as the openssl command successfully created a new private.key file but was not able to move past the "RsaUsingSha alg expects key to be of..." error. I reverted back to 2.1.0 and the CL functionality is working fine. Any ideas?

**Comment by smithrobs** [comment]
@sidsharma27 Hmm. Using 2.2.0-rc2, were you able to get any debug logs? It should be reporting a more specific message in the logs instead of the more general "alg expects key to be of..." exception.

**Comment by sidsharma27** [comment]
With 2.2.0-rc2 - I wasnt thrown that error actually
Thrown the same error as above: TypeInitializationException: The type initializer for 'Nexmo.Api.Configuration' threw an exception.
Nexmo.Api.Request.ApiRequest.GetBaseUriFor(Type component, string url)
I cannot perform **any** actions such Sending an SMS or Starting a verification. 
(tried storing credentials in web.config & appsettings.json but both scenarios threw the same error)

With 2.1.0 I was able to run everything fine but on VoiceCalls it was throwing me "alg expects key to be of" which I am guessing was fixed with 2.2.0-rc2 as mentioned in Issue #22.

**Comment by smithrobs** [comment]
Ah, thank you. I understand. With 2.2.0-rc2, what did the debug log say in regards to attempting to find the config file? At the debug/info level the library should be outputting the file locations and ultimately what credentials are available to the library.

Yes, 2.2.0 addresses #22 by accepting both PKCS#1 and PKCS#8 keys.

**Comment by sidsharma27** [comment]
Updating to VS 17 RC solved the issue. I will go ahead and create a new app with 2.2.0-rc in VS 15 & c/p  the debug log in order find the issue.

**Comment by smithrobs** [comment]
I have made the decision to remove web.config support in v3.0. I was unable to find a good way to make use of IHostingEnvironment outside of the web application scenario.

Closing this issue. I realize it has morphed somewhat over the 3+ months it has been open and there may be a separate issue with 2.2.0 configuration. If needed, feel free to file a new issue so we can have a clean slate to work with. Thanks!
